containers/DvmState.ts

```diff
@@ -0,0 +1,99 @@
+import { createContainer } from "unstated-next";
+import { useState, useEffect } from "react";
+import { utils } from "ethers";
+
+import Connection from "./Connection";
+import DvmContract from "./DvmContract";
+import EmpState from "./EmpState";
+import EmpAddress from "./EmpAddress";
+
+const { formatUnits: fromWei } = utils;
+
+interface ContractState {
+  hasEmpPrice: boolean | null;
+  resolvedPrice: number | null;
+}
+
+const initState = {
+  hasEmpPrice: null,
+  resolvedPrice: null,
+};
+
+// Taken from Voting.sol implementation.
+enum PRICE_REQUEST_STATUSES {
+  NOT_REQUESTED, // Was never requested.
+  ACTIVE, // Is being voted on in the current round.
+  RESOLVED, // Was resolved in a previous round.
+  FUTURE, // Is scheduled to be voted on in a future round.
+}
+
+const useContractState = () => {
+  const { block$ } = Connection.useContainer();
+  const { contract: dvm } = DvmContract.useContainer();
+  const { empState } = EmpState.useContainer();
+  const { priceIdentifier, expirationTimestamp } = empState;
+  const { empAddress } = EmpAddress.useContainer();
+
+  const [state, setState] = useState<ContractState>(initState);
+
+  // get state from EMP
+  const queryState = async () => {
+    if (
+      dvm !== null &&
+      priceIdentifier !== null &&
+      expirationTimestamp !== null
+    ) {
+      const res = await Promise.all([
+        dvm.hasPrice(
+          priceIdentifier.toString(),
+          expirationTimestamp.toNumber(),
+          { from: empAddress }
+        ),
+      ]);
+
+      const hasPrice = res[0] as boolean;
```

Comment:
could you add a comment here to explain exactly the type returned from the `hasPrice` call?

Reply:
It's the same as in `EMPState`, I think its a boolean but with a smart contract call you're never sure I think

```
      const newState: ContractState = {
        expirationTimestamp: res[0] as BigNumber,
        collateralCurrency: res[1] as string, // address
        priceIdentifier: res[2] as Bytes,
        tokenCurrency: res[3] as string, // address
        collateralRequirement: res[4] as BigNumber,
        disputeBondPct: res[5] as BigNumber,
        disputerDisputeRewardPct: res[6] as BigNumber,
        sponsorDisputeRewardPct: res[7] as BigNumber,
        minSponsorTokens: res[8] as BigNumber,
        timerAddress: res[9] as string, // address
        cumulativeFeeMultiplier: res[10] as BigNumber,
        rawTotalPositionCollateral: res[11] as BigNumber,
        totalTokensOutstanding: res[12] as BigNumber,
        liquidationLiveness: res[13] as BigNumber,
        withdrawalLiveness: res[14] as BigNumber,
        currentTime: res[15] as BigNumber,
        isExpired: Number(res[15]) >= Number(res[0]),
        contractState: Number(res[16]),
        finderAddress: res[17] as string, // address
      };

```